lib/net/pool.js

```diff
@@ -3508,6 +3528,139 @@ class Pool extends EventEmitter {
 
     return enabled;
   }
+
+  /**
+   * Selects peers for highBW mode - BIP152
+  */
+
+  async setHighBWPeer(peer) {
+    //before adding peer to the list, checking whether the peer is un-destroyed
+    let peerStatus = false;
+    try {
+      await peer.handlePacket();
+      peerStatus = true;
+    } catch(e) {
+      console.log(e)
+    }
+
+    this.bindPeer(peer);
+
+    if (peerStatus) {
+      if (peer.inbound) {
+        // remove oldest outbound iff there exists atleast 1 outbound peer.
+        if (this.highBWPeers.size() < this.highBWMax && this.highBWPeers.outbound <= 1) {
+          // simply add the new peer to the peer list
+          this.highBWPeers.add(peer);
+            peer.compactModeFrom = common.COMPACT_BLOCKS_HIGHBW;
+            peer.sendCompact(peer.compactModeFrom);
+        }
+
+        else if (this.highBWPeers.size() > this.highBWMax) {
+          // remove the oldest peer
+          this.highBWPeers.remove(this.highBWPeers.head());
+            peer.compactModeFrom = common.COMPACT_BLOCKS_LOWBW;
+            peer.sendCompact(peer.compactModeFrom);
+
+          // add the new peer
+          this.highBWPeers.add(peer);
+            peer.compactModeFrom = common.COMPACT_BLOCKS_HIGHBW;
+            peer.sendCompact(peer.compactModeFrom);
+        }
+        else {
+          const OldestOutboundPeer = this.findOldestHBOutbound();
+
+          // if the peer exists remove that peer
+          if (OldestOutboundPeer != null) {
+            // sending cmpct block mode=0 on removal from highBW Peers list
+            // sending cmpct block mode=1 on addition to highBW Peers list
+            this.highBWPeers.remove(OldestOutboundPeer);
+              peer.compactModeFrom = common.COMPACT_BLOCKS_LOWBW;
+              peer.sendCompact(peer.compactModeFrom);
+
+            // if the size < 3, simply add to the peer list
+            if (this.highBWPeers.size() < this.highBWMax) {
+              this.highBWPeers.add(peer);
+                peer.compactModeFrom = common.COMPACT_BLOCKS_HIGHBW;
+                peer.sendCompact(peer.compactModeFrom);
+            }
+            else if (this.highBWPeers.size() > this.highBWMax) {
+              // if the size > 3, remove the oldest peer and add the new peer
+              this.highBWPeers.remove(this.highBWPeers.head());
+                peer.compactModeFrom = common.COMPACT_BLOCKS_LOWBW;
+                peer.sendCompact(peer.compactModeFrom);
+
+              this.highBWPeers.add(peer);
+                peer.compactModeFrom = common.COMPACT_BLOCKS_HIGHBW;
+                peer.sendCompact(peer.compactModeFrom);
+            }
+          }
+        }
+      }
+
+      if (peer.outbound) {
```

Comment:
Ok I think the logic in this function may be a bit messy, lets try to simplify it:

```
// Simplest case, we have room to add one more
if highBWPeers.size < max
  add new peer and return

// Next simplest case, adding an outbound peer
if peer.outbound
  remove oldest (any) peer
  add new peer and return

// Ok now we know the new peer is inbound
// AND we know we have a full HB peer list

// Easiest: replace oldest inbound with new inbound peer
if highBWPeers.inbound > 0
  remove oldest inbound peer
  add new peer and return
  
// final case: we need to remove an outbound peer, can we though?

// no we can't
if highBWPeers.outbound <= 1
  do nothing and return

// yes we can
remove oldest outbound peer
add new peer and return
```